src/System.Net.Http/tests/UnitTests/HPack/HPackDecoderTest.cs

```diff
@@ -0,0 +1,583 @@
+            .Concat(_headerNameBytes)
+            .ToArray();
+
+        private static readonly byte[] _headerNameHuffman = new byte[] { (byte)(0x80 | _headerNameHuffmanBytes.Length) }
+            .Concat(_headerNameHuffmanBytes)
+            .ToArray();
+
+        private static readonly byte[] _headerValue = new byte[] { (byte)_headerValueBytes.Length }
+            .Concat(_headerValueBytes)
+            .ToArray();
+
+        private static readonly byte[] _headerValueHuffman = new byte[] { (byte)(0x80 | _headerValueHuffmanBytes.Length) }
+            .Concat(_headerValueHuffmanBytes)
+            .ToArray();
+
+        // &        *
+        // 11111000 11111111
+        private static readonly byte[] _huffmanLongPadding = new byte[] { 0x82, 0xf8, 0xff };
+
+        // EOS                              *
+        // 11111111 11111111 11111111 11111111
+        private static readonly byte[] _huffmanEos = new byte[] { 0x84, 0xff, 0xff, 0xff, 0xff };
+
+        private readonly DynamicTable _dynamicTable;
+        private readonly HPackDecoder _decoder;
+
+        private readonly Dictionary<string, string> _decodedHeaders = new Dictionary<string, string>();
+
+        public HPackDecoderTest()
+        {
+            _dynamicTable = new DynamicTable(DynamicTableInitialMaxSize);
+            _decoder = new HPackDecoder(DynamicTableInitialMaxSize, _dynamicTable);
+        }
+
+        void OnHeader(object state, ReadOnlySpan<byte> headerName, ReadOnlySpan<byte> headerValue)
+        {
+            string name = Encoding.ASCII.GetString(headerName);
+            string value = Encoding.ASCII.GetString(headerValue);
+
+            _decodedHeaders[name] = value;
+        }
+
+        [Fact]
+        public void DecodesIndexedHeaderField_StaticTable()
+        {
+            _decoder.Decode(_indexedHeaderStatic, endHeaders: true, onHeader: OnHeader, onHeaderState: null);
+            Assert.Equal("GET", _decodedHeaders[":method"]);
+        }
+
+        [Fact]
+        public void DecodesIndexedHeaderField_DynamicTable()
+        {
+            // Add the header to the dynamic table
+            _dynamicTable.Insert(_headerNameBytes, _headerValueBytes);
+
+            // Index it
+            _decoder.Decode(_indexedHeaderDynamic, endHeaders: true, onHeader: OnHeader, onHeaderState: null);
+            Assert.Equal(_headerValueString, _decodedHeaders[_headerNameString]);
+        }
+
+        [Fact]
+        public void DecodesIndexedHeaderField_OutOfRange_Error()
+        {
+            var exception = Assert.Throws<HPackDecodingException>(() =>
+                _decoder.Decode(_indexedHeaderDynamic, endHeaders: true, onHeader: OnHeader, onHeaderState: null));
+            Assert.Empty(_decodedHeaders);
+        }
+
+        [Fact]
+        public void DecodesLiteralHeaderFieldWithIncrementalIndexing_NewName()
+        {
+            var encoded = _literalHeaderFieldWithIndexingNewName
```

Comment:
All of these `vars` should be the actual type name.  I can see it's an array of something, but it's not clear of what.